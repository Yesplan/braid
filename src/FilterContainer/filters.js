export const hasUnresolvedBlockers = blockers =>
  blockers.some(blocker => !blocker.resolved);

const isBlocked = story => hasUnresolvedBlockers(story.blockers);

export const filterByOwner = (story, ownerIds) => {
  if (ownerIds.length === 0) {
    return true;
  }

  return story.ownerIds.some(id => ownerIds.includes(id));
};

export const filterByEpic = (story, epicIds) => {
  if (epicIds.length === 0) {
    return true;
  }

  return story.labels.some(label => epicIds.includes(label.id));
};

export const filterByType = (story, typeNames) => {
  if (typeNames.length === 0) {
    return true;
  }

  return (
    typeNames.includes(story.storyType) ||
    (typeNames.includes('blocked') && isBlocked(story))
  );
};

export const filterByStoryStates = (story, storyStates) =>
  storyStates.includes(story.currentState);
