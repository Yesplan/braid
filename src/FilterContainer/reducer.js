import { arrayToggle, arrayRotateForward, arrayRotateBackward } from '../utils';

export const toggleOwner = 'TOGGLE_OWNER';
export const clearOwners = 'CLEAR_OWNERS';
export const selectNextOwner = 'SELECT_NEXT_OWNER';
export const selectPrevOwner = 'SELECT_PREV_OWNER';
export const toggleType = 'TOGGLE_TYPE';

export const reducer = (state, action, uniqueOwnerIds) => {
  const { type, payload } = action;

  switch (type) {
    case toggleOwner: {
      return {
        ...state,
        selectedOwners: arrayToggle(state.selectedOwners, payload)
      };
    }

    case clearOwners: {
      return {
        ...state,
        selectedOwners: []
      };
    }

    case selectNextOwner: {
      return {
        ...state,
        selectedOwners: [
          arrayRotateForward(uniqueOwnerIds, state.selectedOwners[0])
        ]
      };
    }

    case selectPrevOwner: {
      return {
        ...state,
        selectedOwners: [
          arrayRotateBackward(uniqueOwnerIds, state.selectedOwners[0])
        ]
      };
    }

    case toggleType: {
      return {
        ...state,
        selectedTypes: arrayToggle(state.selectedTypes, payload)
      };
    }

    default: {
      return { ...state };
    }
  }
};
