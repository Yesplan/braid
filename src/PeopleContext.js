import { createContext } from 'react';

const { Consumer, Provider } = createContext({});

export { Consumer as PeopleConsumer, Provider as PeopleProvider };
