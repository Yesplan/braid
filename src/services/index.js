export { default as window } from './window';
export * from './localStorage';
