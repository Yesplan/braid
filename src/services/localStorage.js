import window from './window';

const localStorage = (function() {
  let listeners = [];

  const triggerListeners = key => {
    listeners
      .filter(listener => !key || key === listener.key)
      .forEach(listener => {
        listener.callBack(window.localStorage.getItem(listener.key));
      });
  };

  window.addEventListener('storage', e => {
    triggerListeners(e.key);
  });

  return {
    addEventListener: (key, callBack) => {
      listeners.push({ key, callBack });
    },

    removeEventListener: (key, callBack) => {
      listeners = listeners.filter(
        listener => !(listener.key === key && listener.callBack === callBack)
      );
    },

    getItem: key => window.localStorage.getItem(key),

    setItem: (key, value) => {
      window.localStorage.setItem(key, value);
      triggerListeners(key);
    },

    removeItem: key => {
      window.localStorage.removeItem(key);
      triggerListeners(key);
    },

    clear: () => {
      window.localStorage.clear();
      triggerListeners(undefined);
    }
  };
})();

export { localStorage };
