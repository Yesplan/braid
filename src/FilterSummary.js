import React, { useContext } from 'react';
import { Initials } from './Initials';
import { Epics } from './Epics';
import { FilterContext } from './FilterContainer';
import { storyTypes } from './storyTypes';

function FilterSummary() {
  const { selectedTypes, selectedOwners, selectedEpics } = useContext(
    FilterContext
  );

  return (
    <div className="tags">
      {selectedOwners.map(id => (
        <span className="tag is-primary" key={id}>
          <Initials id={id} />
        </span>
      ))}

      {selectedEpics.map(id => (
        <span className="tag is-primary" key={id}>
          <Epics id={id} />
        </span>
      ))}

      {storyTypes
        .filter(type => selectedTypes.includes(type.key))
        .map(type => (
          <span
            className={`tag is-uppercase ${type.activeClass}`}
            key={type.key}
          >
            {type.key}
          </span>
        ))}
    </div>
  );
}

export { FilterSummary };
