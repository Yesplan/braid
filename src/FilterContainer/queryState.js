import { window } from '../services';
import { normalizeOwners, normalizeTypes } from './normalize';

const queryString = require('query-string');

const queryStringOptions = {
  arrayFormat: 'comma'
};

function getQueryState(uniqueOwnerIds) {
  const queryState = queryString.parse(
    window.location.search,
    queryStringOptions
  );

  return {
    selectedOwners: normalizeOwners(queryState.selectedOwners, uniqueOwnerIds),
    selectedTypes: normalizeTypes(queryState.selectedTypes)
  };
}

function setQueryState(state) {
  window.history.pushState(
    {},
    document.title,
    `${window.location.pathname}?${queryString.stringify(
      state,
      queryStringOptions
    )}`
  );
}

export { getQueryState, setQueryState };
