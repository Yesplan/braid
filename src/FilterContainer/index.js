import React, { useState, createContext, useEffect, useMemo } from 'react';
import { arrayOf, number, node } from 'prop-types';
import { window } from '../services';
import { filterByOwner, filterByType, filterByStoryStates } from './filters';
import {
  reducer,
  selectNextOwner,
  selectPrevOwner,
  clearOwners
} from './reducer';
import { getQueryState, setQueryState } from './queryState';
import { useKeyup } from '../useKeyup';

const FilterContext = createContext();

const FilterContainer = ({ uniqueOwnerIds, children }) => {
  const [state, setState] = useState(getQueryState(uniqueOwnerIds));

  const dispatch = useMemo(
    () => action => {
      const nextState = reducer(state, action, uniqueOwnerIds);

      setState(nextState);
      setQueryState(nextState);
    },
    [state, uniqueOwnerIds]
  );

  useEffect(() => {
    const syncStates = () => {
      setState(getQueryState(uniqueOwnerIds));
    };

    window.addEventListener('popstate', syncStates);

    return () => {
      window.removeEventListener('popstate', syncStates);
    };
  }, [uniqueOwnerIds]);

  const keyMap = useMemo(
    () => ({
      n: () => dispatch({ type: selectNextOwner }),
      p: () => dispatch({ type: selectPrevOwner }),
      c: () => dispatch({ type: clearOwners })
    }),
    [dispatch]
  );

  useKeyup(keyMap);

  const filter = useMemo(
    () => (storyIds, stories, storyStates) =>
      storyIds
        .map(id => stories[id])
        .filter(
          story =>
            filterByOwner(story, state.selectedOwners) &&
            filterByType(story, state.selectedTypes) &&
            filterByStoryStates(story, storyStates)
        ),
    [state.selectedOwners, state.selectedTypes]
  );

  const value = useMemo(
    () => ({
      ...state,
      uniqueOwnerIds,
      dispatch,
      filter
    }),
    [state, uniqueOwnerIds, dispatch, filter]
  );

  return (
    <FilterContext.Provider value={value}>{children}</FilterContext.Provider>
  );
};

FilterContainer.propTypes = {
  uniqueOwnerIds: arrayOf(number),
  children: node
};

export { FilterContainer, FilterContext };
export { hasUnresolvedBlockers } from './filters';
export {
  toggleOwner,
  clearOwners,
  selectNextOwner,
  selectPrevOwner,
  toggleType
} from './reducer';
