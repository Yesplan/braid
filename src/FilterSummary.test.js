import React from 'react';
import { render } from '@testing-library/react';
import { PeopleContext } from './PeopleContext';
import { FilterContainer } from './FilterContainer';
import { FilterSummary } from './FilterSummary';
import { window } from './services';

jest.mock('./services/window');

describe('FilterSummary', () => {
  it('renders according to snapshot', () => {
    const people = {
      1: { initials: 'AA' },
      2: { initials: 'BB' },
      3: { initials: 'CC' }
    };

    const uniqueOwnerIds = [1, 2, 3];

    window.location.search = 'selectedOwners=1,2&selectedTypes=bug,feature';

    const { container } = render(
      <PeopleContext.Provider value={people}>
        <FilterContainer uniqueOwnerIds={uniqueOwnerIds}>
          <FilterSummary />
        </FilterContainer>
      </PeopleContext.Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
