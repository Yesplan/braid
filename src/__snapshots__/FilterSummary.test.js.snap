// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSummary renders according to snapshot 1`] = `
<div
  class="tags"
>
  <span
    class="tag is-primary"
  >
    <span
      class="is-uppercase"
    >
      AA
    </span>
  </span>
  <span
    class="tag is-primary"
  >
    <span
      class="is-uppercase"
    >
      BB
    </span>
  </span>
  <span
    class="tag is-uppercase is-primary"
  >
    feature
  </span>
  <span
    class="tag is-uppercase is-danger"
  >
    bug
  </span>
</div>
`;
